response cache & replica corruption handling

tbg opened the pull request:
- always update response cache in applyRaftCommand, allowing early returns
- add replicaCorruptionError, bogus handling and test
- anticipate #1400 by not using reply.Header().GoError() as authoritative
  error source and by setting it only in r.processRaftCommand (though
  it is still set in r.executeCmd and below) for now
- deduplicate some logic around the verification of leader leases
- move leadership check from applyRaftCommand to processRaftCommand and add
  a comment

tbg committed:
response cache & replica corruption handling

- always update response cache in applyRaftCommand, allowing early returns
- add replicaCorruptionError, bogus handling and test
- anticipate #1400 by not using reply.Header().GoError() as authoritative
  error source and by setting it only in r.processRaftCommand (though
  it is still set in r.executeCmd and below) for now
- deduplicate some logic around the verification of leader leases
- move leadership check from applyRaftCommand to processRaftCommand and add
  a comment

tamird commented on storage/range.go:
```diff
@@ -387,11 +387,16 @@ func (r *Range) redirectOnOrAcquireLeaderLease(timestamp proto.Timestamp) error
-	l := r.getLease()
-	return originRaftNodeID == l.RaftNodeID && timestamp.Less(l.Expiration)
+// verifyLeaderLease checks whether the requesting replica (by raft node ID)
+// holds the leader lease covering the specified timestamp.
```
not exactly accurate:
```
// verifyLeaderLease checks whether `lease` is held by the replica specified by replicaRaftNodeID
// at the given timestamp.
```

tamird replied:
also mention something about the return values

tschottdorf replied:
verifyLeaderLease is now gone.

tamird commented on storage/range.go:
```diff
@@ -387,11 +387,16 @@ func (r *Range) redirectOnOrAcquireLeaderLease(timestamp proto.Timestamp) error
+		return false, true
+	}
+	held := lease.RaftNodeID == replicaRaftNodeID
+	expired := !timestamp.Less(lease.Expiration)
```
the lease is expired when at expiration?

tschottdorf replied:
no, but it doesn't really matter and the refactor you suggested above (which I'm in the middle of) suggests that it's better (more comfortable) to exclude the endpoint, so I'll likely make that change.

tschottdorf replied:
endpoint is now excluded, see `Lease.Covers(ts)` in `proto/data.go`.

tamird commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
 	r.Unlock()
 
 	args := raftCmd.Cmd.GetValue().(proto.Request)
+	origin := proto.RaftNodeID(raftCmd.OriginNodeID)
```
nit: rename to originNodeID

tschottdorf replied:
done.

tamird commented on storage/range.go:
```diff
@@ -387,11 +387,16 @@ func (r *Range) redirectOnOrAcquireLeaderLease(timestamp proto.Timestamp) error
-	return originRaftNodeID == l.RaftNodeID && timestamp.Less(l.Expiration)
+// verifyLeaderLease checks whether the requesting replica (by raft node ID)
+// holds the leader lease covering the specified timestamp.
+func verifyLeaderLease(lease *proto.Lease, replicaRaftNodeID proto.RaftNodeID, timestamp proto.Timestamp) (bool, bool) {
```
This method is pretty weird. It has two callers:
- `processRaftCommand`, which only cares if the lease is live for the supplied node, and
- `HasLeaderLease`, which has a few callers:
  - `redirectOnOrAcquireLeaderLease, which only cares if the lease is live for another range,
  - `shouldQueue` which only cares if the lease is live for **any** range,
  - `updateStoreStatus`, which only checks if the lease is live for **any** range, but seems like it should check if the lease is live for the caller.

I think this API is super clumsy. Can we just have two methods?
`func (l *lease) Live(timestamp Timestamp) bool` returns true if the lease is unexpired.
`func (r *Range) HasLiveLeaderLease(timestamp proto.Timestamp) bool` which returns true if the lease is live and held by this range.

tschottdorf replied:
I'll give this (or something like this) a try, though `redirectOnOrAcquireLeaderLease` actually needs both the information about the lease holder and the expiry separately.

tschottdorf replied:
the check in `updateStoreStatus` looks ok. The thinking is that if someone holds the leader lease for the range, then that range (as a Raft group) is doing well.

tamird replied:
ah, then the comment should be updated (currently reads `// If the range has the leader lease, then it's available.`)

bdarnell replied:
`func (r *Range) HasLiveLeaderLease(timestamp proto.Timestamp) bool` can't work because we often need to verify whether a specified replica has the lease at a time.  When a raft command is applied all replicas verify that the origin replica held the lease at the time the command was proposed.  We still need to be able to pass in the node id. 

"Live" is confusing here since we're passing in a timestamp and that timestamp will often be in the past.

I would try splitting this function into `verifyLeaderLease(*proto.Lease, proto.RaftNodeID, proto.Timestamp) bool` (returning the first result of this function) and `isLeaseAvailable(*proto.Lease, proto.Timestamp)` (returning the second).

tschottdorf replied:
addressed all of the above.

tamird commented on storage/range.go:
```diff
@@ -805,10 +841,19 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
+			// raft commands so that every replica maintains the same responses
+			// to continue request idempotence, even if leadership changes.
+			if pErr := r.respCache.PutResponse(args.Header().CmdID, reply); pErr != nil {
+				rErr = newReplicaCorruptionError("could not put to response cache", pErr)
```
this can potentially overwrite the previously-set `rErr`. is that intended?

tschottdorf replied:
adressed below.

tamird commented on storage/range.go:
```diff
@@ -805,10 +841,19 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
+			// is a write method. This must be done as part of the execution of
+			// raft commands so that every replica maintains the same responses
+			// to continue request idempotence, even if leadership changes.
+			if pErr := r.respCache.PutResponse(args.Header().CmdID, reply); pErr != nil {
```
nit: you have a similar pattern on L844, which calls its error variable `err`. Maybe be consistent and rename that one to `sErr` or something?

tschottdorf replied:
done.

tschottdorf replied:
done.

bdarnell replied:
Why not just `err` for both?

tschottdorf replied:
done.

tamird commented on storage/range.go:
```diff
@@ -819,21 +864,9 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
-			reply.Header().SetGoError(err)
 			return err
+		} else if ok && err != nil {
+			return newReplicaCorruptionError("could not read from response cache", err)
```
this is going to get overwritten in the defer, which is super surprising >.<

tschottdorf replied:
yes, that shouldn't happen. I'll make sure the first replicaCorruptionError seals the deal.

tschottdorf replied:
moved this one up above the `defer()` (which is a good place for it).
For the rest, the deferred replica corruption errors now always chain with the originally returned value, so they don't lose information (we might get two replica corruption errors chained together, but that's still helpful information).

tschottdorf replied:
done.

tamird commented on storage/range.go:
```diff
@@ -805,10 +841,19 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 		if !committed {
 			// We didn't commit the batch, but advance the last applied index nonetheless.
 			if err := setAppliedIndex(r.rm.Engine(), r.Desc().RaftID, index); err != nil {
```
this exactly happens in the normal function body (outside the defer) why?

tschottdorf replied:
in this version of the code, it can move back to the main body. Will do that in the next update.

tschottdorf replied:
sorry, I was mistaken. It's most elegant to do this up here, we want to do this rain or shine and there are a bunch of points at which the main function returns (most prominently if `executeCmd` comes back with an error).

tamird commented on storage/range_test.go:
```diff
@@ -2284,6 +2284,39 @@ func TestAppliedIndex(t *testing.T) {
+		proto.Call{Args: args, Reply: reply}, true /* wait */)
+
+	if err != nil && strings.Contains(err.Error(), "replica corruption (processed=true)") {
+		return
```
i think i'd prefer the inverse check to avoid an explicit return in a test

tschottdorf replied:
done.

tamird commented on storage/range.go:
```diff
@@ -1035,6 +1058,47 @@ func (r *Range) handleSkippedIntents(args proto.Request, intents []proto.Intent)
 
+type replicaCorruptionError struct {
+	error
+	msg       string
```
this much state is unusual, right? can you add a comment explaining what it's all for?

tschottdorf replied:
there's now a smaller, likely self-explanatory state.

bdarnell commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
+		// This error case is somewhat special: Any command that makes it into
+		// Raft should have the leader lease held by the proposer. The only
+		// case that violates this assumption is that of the proposer having a
+		// corrupted internal state and erroneously assuming leadership. Then,
```
I think raft-level forwarding can also violate this assumption. We've talked about disabling that but haven't yet.

tschottdorf replied:
Are you sure? The "proposer" here is the replica which put the command into the Raft black box (see `rng.proposeRaftCommand`), it doesn't matter which replica turns out to be the Raft leader committing it.

I remember now why this is really here though, it's because checking the leader lease once isn't enough: it doesn't remain the same (for a fixed timestamp) as new leader leases are passed out, so a command that spends a long time in Raft may pop out and find it's been proposed by a replica whose leadership has by then ceded to be.
I would argue however that it's fine to execute such commands without the additional check (unless we're worried about corrupted replicas) because a command having been proposed means that the leader lease was held at propose time, and any later lease holder will have a timestamp cache compatible with any such read from a previous lease period. Writes aren't really affected by leases as much, so that leaves admin commands, which are ok too. There might still be some subtleties so I'd leave this in for now.

tschottdorf replied:
I take the second part back, was thinking about the timestamp cache too much and not about the rest. Writes need to see all previous reads, and those happen on the leader. So without this check, the old leader could propose a write which makes it out of Raft only after the key affected by the write has seen a read with a higher timestamp that only the new leader knows about.

Will update the comment accordingly.

spencerkimball replied:
I think we still need the forwarding. Let's say there's a raft leader but no leader lease. In that case, if a follower replica is going to acquire the leader lease, it needs to be able to forward at the raft level. The impetus to disable forwarding was mainly a product of trying to make leader leases and raft leaders always the same replica.

bdarnell commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
+		// the command should certainly not be executed by the replicas, but
+		// the error must not enter the response cache (which is updated in
+		// applyRaftCommand below) since it is going to be retried with exactly
+		// the same timestamp. On retry, it may be proposed by the real leader
```
s/timestamp/ClientCmdID/

tschottdorf replied:
done.

bdarnell commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
+		err = r.newNotLeaderError()
+	} else {
+		err = r.maybeSetCorrupt(
+			r.applyRaftCommand(ctx, index, origin, args, reply),
```
It's a little weird to see applyRaftCommand (the central action of this function) buried as a parameter to maybeSetCorrupt. I'd pull this out into separate statements (and maybe change `maybeSetCorrupt` to `if err != nil { err = r.setCorrupt() }`)

tschottdorf replied:
moved intent handling up from `applyRaftCommand`, this changed as a result. See if you like it more now.

spencerkimball replied:
+1

bdarnell commented on storage/range.go:
```diff
@@ -819,21 +864,9 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
-		}
-	}
-
-	// Verify the leader lease is held; Note that we don't require the
```
The old code checked the response cache before the leader lease on purpose: if there is an entry in the response cache then this is a retry and the lease was valid at the time the command was executed, so we can return the result. With this change it is possible for a client to submit a command, lose its connection (so it doesn't see that the command succeed), then retry and see an error. Of course, the NotLeaderError is retryable so the client will just try again on the new leader and should eventually get through, so I think all this really saves us is an additional retry on rare cases that are already failing and retrying.

tschottdorf replied:
It's moved back up due to another comment by @tamird.

tbg committed:
rework lease API

as suggested by @tamird.

added OwnedBy and Covers to Lease itself,
which obviated the need for Range.HasLeaderLease.

tbg committed:
pass lease to newNotLeaderError

in all situations in which we create this error
we already have a lease, and this possibly avoids
some (likely harmless) races.

tschottdorf commented:
updated, PTAL.

tamird commented on storage/range.go:
```diff
@@ -1035,6 +1061,76 @@ func (r *Range) handleSkippedIntents(args proto.Request, intents []proto.Intent)
+// which puts its integrity at risk.
+type replicaCorruptionError struct {
+	error
+	processed bool // for now, required for testing
```
can you expand this comment to indicate that "processed" means this error has been "handled"?

tschottdorf replied:
done.

tamird commented:
LGTM thanks for the refactoring!

tschottdorf commented:
@tamird you're welcome.

@bdarnell I'll wait for your LGTM as well.

bdarnell commented on storage/range.go:
```diff
@@ -819,23 +852,37 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
+
+	defer func() {
+		if !committed {
+			// We didn't commit the batch, but advance the last applied index nonetheless.
```
This should come before the read from the response cache. We should advance the applied index even if we "applied" the command by reading a cached response. In fact, we also need to advance the applied index if we're returning a notLeaderError (since the retried command will be at a new log index)

tschottdorf replied:
Moved the `defer` to the top and pulled in the `notLeaderError` check from `processRaftCommand` to `applyRaftCommand`. In turn removed the last commit (passing `originNodeID` into `applyRaftCommand` and dealing with intents in it).

bdarnell commented on storage/range.go:
```diff
@@ -845,27 +892,33 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
+	}
+
+	// If the execution of the command wasn't successful, stop here.
+	if err != nil {
```
This can be moved up to immediately after `r.executeCmd`. There's no reason to carry an unexamined error this far. (and then the `setAppliedIndex` block can just use the name `err` instead of dancing around with `err` and `e`).

tschottdorf replied:
done.

bdarnell commented on storage/range.go:
```diff
@@ -819,23 +852,37 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID
 		}
-	}
+		if proto.IsWrite(args) {
+			// No matter the result, add result to the response cache if this
```
The error is a part of the response and must be saved in the response cache. We must do `Reply.Header().SetGoError(rError)` here.

tschottdorf replied:
done.

tbg committed:
fix order

spencerkimball commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
+	// lease when trying to grant the leader lease!
+	held, expired := verifyLeaderLease(r.getLease(), origin, args.Header().Timestamp)
+	if (!held || expired) && args.Method() != proto.InternalLeaderLease {
+		// This error case is somewhat special: Any command that makes it into
```
This is a big comment for a byzantine corruption failure. If the only way this can happen is through a corruption of internal state, then it's simply not something we should be working to handle properly. If we did that for every place in the code where an adversarial internal corruption could occur, we would just end up with an endlessly growing file of special case checks and error handling. If this should never happen, then assert that with a panic here.

bdarnell replied:
It's not just byzantine; the second part of the comment about a change in leadership is the important part. I'd just drop the part about a corrupted replica.

tschottdorf replied:
dropped.

spencerkimball commented on storage/range.go:
```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
-// The caller needs to hold the range lock.
-func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNodeID proto.RaftNodeID, args proto.Request, reply proto.Response) error {
+// applyRaftCommand applies a raft command from the replicated log to the
+// underlying state machine (i.e. the engine). The caller needs to hold the
```
Does the caller in fact need to hold the range lock? That doesn't appear to be true.

tschottdorf replied:
you're right. Dropping the comment.

spencerkimball commented on storage/range.go:
How about have maybeSetCorrupt write a bit in the local range descriptor for this range to mark it as corrupt, then call `r.rm.RemoveRange(r)`, and simply ensure that when we scan ranges in the store at startup, we skip any marked corrupt? That will permanently remove the range without immediately deleting it.

spencerkimball commented:
Have we gotten any replica corruptions yet?

spencerkimball commented on storage/range.go:
```diff
@@ -1035,6 +1069,78 @@ func (r *Range) handleSkippedIntents(args proto.Request, intents []proto.Intent)
 	}
 }
 
+type chainedError struct {
```
This feels like something for the util package.

tschottdorf replied:
added a TODO. Convenient to do this when actually doing something with `setCorrupt`.

tschottdorf commented on storage/range.go:
I didn't want to get into actual handling yet for this PR, it's already getting a bit out of hand.

spencerkimball commented on storage/range.go:
```diff
@@ -760,57 +747,103 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
+	var intents []proto.Intent
+	if args.Method() != proto.InternalLeaderLease &&
+		(!lease.OwnedBy(originNode) || !lease.Covers(args.Header().Timestamp)) {
+		// This error case is somewhat special: Any command that makes it into
```
The comment  is so big for a relatively straightforward situation. How about we rephrase as:

    // If the lease was lost by this replica during the time the command
    // took from Raft proposal until now, redirect to the new lease holder.
    // This is strictly necessary because the new lease holder may have
    // already served a read which conflicts with this write operation.

tschottdorf replied:
Given that this is going to be out in the weeds for most readers, I've kept the comment a bit more in-depth but significantly trimmed from the version in this diff.

tbg committed:
clean up responsibilities of {process,apply}RaftCommand

address @bdarnell's comments, in particular

- proposeRaftCommand will *always* call applyRaftCommand
- applyRaftCommand
  - checks leader lease of proposer (with correct response cache handling)
  - resolves intents asynchronously
  - **always** advances applied index
  - if anything goes wrong: replicaCorruption

tschottdorf commented:
@spencerkimball no replica corruptions yet. A good way to get some (both for tests and, presumably, real life) will be #837.

tschottdorf commented:
@bdarnell I believe I've addressed your points. PTAL.

spencerkimball commented:
LGTM

bdarnell commented:
LGTM
